### PR TITLE
Preserve Union type annotations during serialization

### PR DESCRIPTION
**Pull request by ansley**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #58911
* #58673
* #57014
* __->__ #57013
* #57012
* #57011
* #53180

Differential Revision: [D28572813](https://our.internmc.facebook.com/intern/diff/D28572813)

### CODE REVIEW

**Comment by ansley** [comment]
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28572813).

**Comment by ansley** [comment]
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28572813).

**Comment by ansley** [comment]
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28572813).

**Comment by ansley** [comment]
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28572813).

**Comment by ansley** [comment]
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28572813).

**Comment by ansley** [comment]
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28572813).

**Comment by ansley** [comment]
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28572813).

**Comment by ansley** [comment]
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28572813).

**Comment by ansley** [comment]
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28572813).

**Comment by ansley** [comment]
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28572813).

**Comment by ansley** [comment]
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28572813).

**Comment by ansley** [comment]
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28572813).

**Comment by ansley** [comment]
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28572813).

**Comment by ansley** [comment]
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28572813).

**Comment by ansley** [comment]
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28572813).

**Comment by ansley** [comment]
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28572813).

**Comment by ansley** [comment]
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28572813).

**Comment by ansley** [comment]
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28572813).

**Comment by ansley** [comment]
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28572813).

**Review by eellison — COMMENTED** [review]
@gmagogsfm mind reviewing this one ?

**Comment by gmagogsfm** [comment]
> @gmagogsfm mind reviewing this one ?

I will get it reviewed today

File: torch/csrc/jit/serialization/python_print.cpp [hunk 1]
```diff
@@ -510,13 +510,36 @@ struct PythonPrintImpl {
...
     printValueList(body_, rhs);
     body_ << "\n";
   }
 
   bool requiresAnnotation(Value* lhs, Value* rhs) {
-    return *lhs->type() != *rhs->type();
+    if (lhs->type()->kind() == UnionType::Kind ||
+        lhs->type()->kind() == OptionalType::Kind) {
+      if (!annotated_unions_.count(lhs)) {
+        annotated_unions_.insert(lhs);
```
Comment (gmagogsfm) [inline torch/csrc/jit/serialization/python_print.cpp:535]: Optional Nit:

`insert` will actually return a pair of values, second of which will tell you whether an insertion actually took place or not. If you use that value, you can save one lookup in `count` and logic will be simpler.

File: torch/csrc/jit/serialization/python_print.cpp [hunk 2]
```diff
@@ -510,13 +510,36 @@ struct PythonPrintImpl {
...
     body_ << " = ";
+    // or if value is being assigned to something of a union type
     printValueList(body_, rhs);
     body_ << "\n";
   }
 
   bool requiresAnnotation(Value* lhs, Value* rhs) {
-    return *lhs->type() != *rhs->type();
+    if (lhs->type()->kind() == UnionType::Kind ||
+        lhs->type()->kind() == OptionalType::Kind) {
```
Comment (gmagogsfm) [inline torch/csrc/jit/serialization/python_print.cpp:533]: Good catch! I wonder how this never caught on fire for `OptionalType`

File: torch/csrc/jit/serialization/python_print.cpp [hunk 3]
```diff
@@ -510,13 +510,36 @@ struct PythonPrintImpl {
     }
     indent();
     printValueList(body_, lhs);
+    // We need to preserve Union/Optional type annotations, but only if
+    // we're not assigning values as part of a tuple unpacking statement
```
Comment (gmagogsfm) [inline torch/csrc/jit/serialization/python_print.cpp:514]: Does this mean unpacking tuple into a `union`-typed variable is not supported now? If so, let's add an explicit error and a `TODO` for future fix.

Comment (ansley) [reply torch/csrc/jit/serialization/python_print.cpp:514]: Can you give an example? We only add annotations on the first assignment of a variable, and I can't think of a case in which some lhs could be a `Union` without an annotation. (Obviously, having two Union annotations on the same line is a SyntaxError!) Thank you!

Comment (gmagogsfm) [reply torch/csrc/jit/serialization/python_print.cpp:514]: Thanks for clarification, I didn't know about this Python limitation before.

**Review by gmagogsfm — APPROVED** [review]

**Comment by ansley** [comment]
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28572813).

**Comment by ansley** [comment]
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28572813).

**Comment by ansley** [comment]
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28572813).

**Comment by ansley** [comment]
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28572813).

**Pull request closed without merging**